larpandoracontent/LArMonitoring/VisualParticleMonitoringAlgorithm.cc

```diff
@@ -0,0 +1,428 @@
+            if (!uTrackHits.empty())
+                PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &uTrackHits, "u_" + suffix, BLUE));
+            if (!vTrackHits.empty())
+                PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &vTrackHits, "v_" + suffix, BLUE));
+            if (!wTrackHits.empty())
+                PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &wTrackHits, "w_" + suffix, BLUE));
+        }
+        else
+        {
+            std::string suffix{std::to_string(pfoIdx) + "_S"};
+            if (!uShowerHits.empty())
+                PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &uShowerHits, "u_" + suffix, RED));
+            if (!vShowerHits.empty())
+                PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &vShowerHits, "v_" + suffix, RED));
+            if (!wShowerHits.empty())
+                PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &wShowerHits, "w_" + suffix, RED));
+        }
+        ++pfoIdx;
+    }
+
+    PANDORA_MONITORING_API(ViewEvent(this->GetPandora()));
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void VisualParticleMonitoringAlgorithm::MakeSelection(const MCParticleList *pMCList, const CaloHitList *pCaloHitList,
+    LArMCParticleHelper::MCContributionMap &mcMap)
+{
+    // Default reconstructability criteria are very liberal to allow for unfolded hierarchy
+    LArMCParticleHelper::PrimaryParameters parameters;
+    parameters.m_minPrimaryGoodHits = 2;
+    parameters.m_minHitsForGoodView = 1;
+    parameters.m_maxPhotonPropagation = std::numeric_limits<float>::max();
+    parameters.m_minHitSharingFraction = 0;
+    parameters.m_foldBackHierarchy = false;
+
+    if (!m_isTestBeam)
+    {
+        LArMCParticleHelper::SelectReconstructableMCParticles(pMCList, pCaloHitList, parameters, LArMCParticleHelper::IsBeamNeutrinoFinalState, mcMap);
+        LArMCParticleHelper::SelectReconstructableMCParticles(pMCList, pCaloHitList, parameters, LArMCParticleHelper::IsCosmicRay, mcMap);
+    }
+    else
+    {
+        LArMCParticleHelper::SelectReconstructableMCParticles(pMCList, pCaloHitList, parameters, LArMCParticleHelper::IsBeamParticle, mcMap);
+        LArMCParticleHelper::SelectReconstructableMCParticles(pMCList, pCaloHitList, parameters, LArMCParticleHelper::IsCosmicRay, mcMap);
+    }
+}
+#endif // MONITORING
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode VisualParticleMonitoringAlgorithm::ReadSettings(const TiXmlHandle xmlHandle)
+{
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle, "CaloHitListName", m_caloHitListName));
+    if (m_caloHitListName.empty())
+        m_caloHitListName = "CaloHitList2D";
```

Comment:
This works OK. In other, similar cases, we've assigned these string values to the member variables in the constructor, and then allowed a STATUS_CODE_NOT_FOUND to slide by.

Reply:
I've preferred this method here, because it covers the case where the tag is supplied but empty. Happy to adopt the constructor-based approach if that's preferred.

Reply:
Your call 😄